src/main.c

```diff
@@ -67,8 +87,29 @@ void gpsdo(void)
     HAL_TIM_Base_Start(&htim3);
     HAL_TIM_Encoder_Start(&htim3, TIM_CHANNEL_ALL);
 
+    int contrast_pwm = 0;
+    GPIO_PinState contrast_pin_state;
     while (1) {
+        uint32_t now = HAL_GetTick();
+        if(pps_out_up && now-last_pps_out >= PPS_PULSE_WIDTH)
+        {
+            HAL_GPIO_WritePin(PPS_OUTPUT_GPIO_Port, PPS_OUTPUT_Pin, 0);
+            pps_out_up = false;
+        }
+        if (now >= VCO_ADJUSTMENT_DELAY) {
+            // Start adjusting the VCO after some time
+            frequency_allow_adjustment(true);
+        }
+
         gps_read();
         menu_run();
+
+        contrast_pin_state = (contrast_pwm < (contrast)) ? GPIO_PIN_RESET : GPIO_PIN_SET;
+        HAL_GPIO_WritePin(LCD_CONTRAST_GPIO_Port, LCD_CONTRAST_Pin, contrast_pin_state);
```

Comment:
When I was experimenting with this myself I noticed that the contrast pin has timer alternative function, so this is possible to do with hardware PWM. Any reason you are doing it "manually"?

Reply:
Yes it would be better!
I was just not experienced enough in STM32 development to do that when I started working on this and went the easier way for me!
I should probably change this to using a timer (also I don't have STM32 Cube Programmer installed so I can't update and  regenerate HAL files).